### ISSUE TITLE
Error with 4-spin interaction

### ISSUE THREAD

**Issue by loudesplat** [bertdupe/Matjes#74]
The current version crashes if the 4-spin interaction is set with the following error:
```
Start setting Hamiltonian: 4-spin interaction
At line 313 of file /Matjes/src/setup/neighbor_type.f90
Fortran runtime error: Attempting to allocate already allocated variable 'diff_vec'
```
MWE:
```
simulation magnet-dynamics

M_4spin
1 -1.05d-3 #atom type 1, amplitude in eV


atomtypes 2
Fe  2.7 0.0 0.0 F 
Ir  0.0 0.0 0.0 F

atoms 2
Fe 0.0  0.0   0.0
Ir 0.5 0.5 -0.7


# parameters of the lattice
Periodic_log .T. .T. .F.
Nsize 15 15 1
alat 1.0 1.0 1.0
lattice
0.5 -0.86602540378 0.0
0.5 0.86602540378 0.0
0.0 0.0 1.0


#input for the dynamic
integration 2  (1: simple integration 2:Heun  3: SIB ) 
timestep 1 (fs)
duration 1 (in units of dt)
damping 0.5
```

**Comment by patrick432** [comment]
I cannot reproduce this Error with the master version e5740e130279bda8e404ef1172edb36fb8708684. 
Which version are you using?

For me it works without problems and yields an energy of 12.6meV which is not improbable as the 12-fold of the input interaction.

**Comment by loudesplat** [comment]
I am also using the latest version in master. It seems the issue only arises when compiling with the debug option. Weird?

**Comment by patrick432** [comment]
I also can not reproduce this error using the debug options.
Looking at the code I guess it is an issue with intent(out) with optinal allocatable (which admittedly sound adventurous). I changed to array to intent(inout) and made sure that is gets deallocated if allocated previously.

I hope the references issue fixes the issue.

**Comment by loudesplat** [comment]
I tried with your fix and it now segfaults if I compile with debug, but still works fine with release version...

**Comment by patrick432** [comment]
Hm, does it segfault at another position in the code now or still at the same one?
It could indicate some more serious problem, but since I can't repoduce it with debug finding an error is rather difficult.
Your output might help.

**Comment by loudesplat** [comment]
It segfaults at the same position, right after trying to set the 4-spin. Output is attached.
[out.txt](https://github.com/bertdupe/Matjes/files/7528977/out.txt)

**Comment by patrick432** [comment]
Very weird, now it crashed at the line

        if(allocated(diff_vec)) deallocate(diff_vec)

which is is rather surprising as I though that statement should be rather save.
I would be curious if it crashes at the allocated or the deallocate, in any case the diff_vec seems to be in a very weird state.
One could try to get around it with another local allocatable so that an assigment to diff_vec only happens in the end directly with the "present" check. 
It might be the sympton of a memory corruption, but I somewhat doubt that. After running with debuggers I didn't find something immediately susptious.
However, I can't really play around with that as it works fine with my compiler. 
Also one could try to play with the compiler flags to find out which flag actually causes the segfault compared to the release version.

In any case, I doubt that it is a sympton of an error which could lead to wrong results, maybe it is rather an obscure compiler error present in your compiler version.

**Comment by loudesplat** [comment]
I initially tried removing the optional compiler flags but somehow these:   
  CPP_BLAS
  CPP_EIGEN
  CPP_FFTW3
  CPP_MKL
 are always displayed in the command line output, I can add  or remove other flags so I'm editing the correct init.config file but when i remove them these are still apparently being used (I recompiled from scratch with the sh file). I don't know if I'm missing something there, or if they're not actually optional for some reason.

Anywho I tried running Matjes with gdb and I get this.
![matjes151121](https://user-images.githubusercontent.com/37928020/141759312-69f10e2d-abf1-4616-80aa-d5bd59caf42e.png)
:

**Comment by patrick432** [comment]
Those flags are optional, but during the installation cmake searches for the libraries and adds the flags if they are found.
Those can be disables using eg. set(USE_EIGEN   FALSE) in the config.cmake as described [here](https://github.com/bertdupe/Matjes/wiki/Installation#feature-selection).

I don't entirely get the gdb output, but is seems again that the diff_vec has some size ridiculous size where in fact it should not be allocated at all.
I guess if one really wants to get rid of that one could remove the optional and work around that.
Otherwise I don't really see a good and easy fix.

**Comment by bertdupe** [comment]
problem with the allocation of the table of distances in the routine - the subroutine had as an input distances(1) and was declared as distances(size(neigh)) which could be >1. Since there was no interface, it was mixing memory and had the wrong allocation for the next variable on the same line.
corrected 24.11.2022
